### ISSUE TITLE
About the different initial results.

### ISSUE THREAD

**Issue by ChrisChenYh** [dqj5182/CONTHO_RELEASE#11]
Hi, sorry to bother you again. 
I'm now trying to train the intial reconstruction network. The structure of the network is shown below. 
<img width="614" alt="image" src="https://github.com/user-attachments/assets/6c563a06-438f-4fb0-89fb-96e0518489a4">
And I've tried two different loss functions:
![image](https://github.com/user-attachments/assets/8dc33019-2d0d-4392-81d3-2c28f8feb6f8)
![image](https://github.com/user-attachments/assets/2eb424ea-d742-46aa-9cec-785e51b17a64)
I keep all settings as the same as your original code. I get the final results as below, which are quite different from the result as you shown in Table 2. in the paper.
![image](https://github.com/user-attachments/assets/aee50fce-d047-4e81-ae03-cac9b33ce30d)
![image](https://github.com/user-attachments/assets/1803a0e7-d424-458f-95ee-0775c350ddc0)
I wonder if i'm missing some details or having some misunderstandings. Cloud you offer me some advice.

**Comment by dqj5182** [comment]
No problem!

As far as I remember, the initial reconstruction results is not a result for model trained only for initial reconstruction.
It is the evaluation of the initial reconstructed 3D human and object from final CONTHO model.
So, the initial reconstruction results from Table 2 and your result may differ due to this reason.

But, I think the result seems to be quite worse than my previous experimental results.
Could you re-train the original CONTHO model but just detach the outputs of initial reconstruction, so that initial reconstruction is separately learned with refinement module?

**Comment by ChrisChenYh** [comment]
I notice that you have detach the output of initial reconstruction in the original CONTHO.
<img width="1140" alt="image" src="https://github.com/user-attachments/assets/df0ff435-8791-454b-8674-192b683d646f">
So I just load the weights I trained before, output the initial reconstruction results, and evaluate them. However, the results still seem to be quite bad, though the final results seems good. 
<img width="397" alt="image" src="https://github.com/user-attachments/assets/8ffc48ed-a01a-4c5c-890b-3fb54e88a7a5">
My operations are shown as below:
(1)output the initial smpl mesh verts, and initial obj pose rotmat.
<img width="1136" alt="image" src="https://github.com/user-attachments/assets/cc42d552-f54b-4bba-9ffc-bfc25e9f8d0f">
<img width="443" alt="image" src="https://github.com/user-attachments/assets/17d2fcb8-df3d-43d2-93ea-2a4d42ea6f55">

(2) Add evaluation code in Tester.evaluation
<img width="1156" alt="image" src="https://github.com/user-attachments/assets/e902c973-6139-4a93-ae13-0b35789d3e97">

<img width="1364" alt="image" src="https://github.com/user-attachments/assets/d19ce413-69d1-48fb-b039-97e598472292">

**Comment by dqj5182** [comment]
Sorry. I'll get back to you after CVPR submission deadline (about two weeks later).
But, after some glimpse, the performance seems lower than I expect.

**Comment by ChrisChenYh** [comment]
Okay, thank you very much. Just at your convenience. Good luck with CVPR！

**Comment by lnykyks** [comment]
Another problem related to the initial pose is that when visualizing the human body in the initial pose, the shape of the human body is very strange. For example, the rotation of the right foot joint makes the right heel overlap with the left thigh.

**Comment by dqj5182** [comment]
@ChrisChenYh 
Sorry for late reply and thank you for your patience.

Please follow below to get evaluation of initial reconstruction by CONTHO.
For [L196](https://github.com/dqj5182/CONTHO_RELEASE/blob/62659cad8d159bcf00e7b5f40bb609a199b00cf7/lib/core/base.py#L196)-[L206](https://github.com/dqj5182/CONTHO_RELEASE/blob/62659cad8d159bcf00e7b5f40bb609a199b00cf7/lib/core/base.py#L206) of base.py, change as follows:
`
init_smpl_mesh_cam = outputs['init_smpl_verts'][i]
            smpl_joint_cam = smplh.joint_regressor @ smpl_mesh_cam
            
            smpl_mesh_cam = smpl_mesh_cam - smpl_joint_cam[smplh.root_joint_idx]
            pred_smpl_mesh_cam.append(init_smpl_mesh_cam)

            obj_name = meta_info['obj_name'][i]
            obj_mesh = obj_dict[obj_name].load_template()
            obj_faces.append(np.array(obj_mesh.faces))

            obj_mesh_cam = obj_dict.transform_object(obj_mesh.vertices, outputs['initial_obj_pose'][i], outputs['initial_obj_trans'][i])
`

and for [L46](https://github.com/dqj5182/CONTHO_RELEASE/blob/62659cad8d159bcf00e7b5f40bb609a199b00cf7/lib/models/model.py#L46)-[L89](https://github.com/dqj5182/CONTHO_RELEASE/blob/62659cad8d159bcf00e7b5f40bb609a199b00cf7/lib/models/model.py#L89) of model.py, change as follows:
`
        human_verts, human_joints = outputs.vertices - outputs.joints[:, [smplh.root_joint_idx]], outputs.joints - outputs.joints[:, [smplh.root_joint_idx]]

        init_human_verts = human_verts.clone()

        human_verts = smplh.downsample(human_verts)

        obj_ids = inputs['obj_id'].reshape(-1).tolist()
        obj_pose_ = rot6d_to_rotmat(obj_pose).reshape(batch_size, 3, 3)
        obj_verts = torch.zeros((batch_size, 64, 3))
        for i, idx in enumerate(obj_ids):
            obj_verts[i] = torch.tensor(obj_dict[idx].load_verts())
        obj_verts = torch.matmul(torch.tensor(obj_verts).to(obj_pose_.device), obj_pose_.transpose(1,2)) + obj_trans[:,None]

        init_obj_verts = obj_verts.clone()
        init_obj_pose, init_obj_trans = self.kps_to_Rt(init_obj_verts, obj_ids)

        # Projection & Feature sampling
        human_joints_proj = self.projection(human_joints, cam_trans)
        human_verts_proj = self.projection(human_verts, cam_trans)
        obj_verts_proj = self.projection(obj_verts, cam_trans)

        smpl_coord_xy = (human_verts_proj[...,:2].detach() + 0.5) * img_feat.shape[-1]
        human_feat = sample_joint_features(img_feat, smpl_coord_xy) 
        obj_coord_xy = (obj_verts_proj[...,:2].detach() + 0.5) * img_feat.shape[-1]
        obj_feat = sample_joint_features(img_feat, obj_coord_xy)

        # ContactFormer
        human_tokens, object_tokens, h_contacts, o_contacts = self.contactformer(human_verts.detach(), obj_verts.detach(), human_feat, obj_feat) 
        
        # CRFormer
        refined_human_verts, refined_obj_verts = self.crformer(human_tokens, object_tokens, h_contacts, o_contacts) 
    
        # Get final meshes 
        refined_human_verts = smplh.upsample(refined_human_verts)
        refined_obj_pose, refined_obj_trans = self.kps_to_Rt(refined_obj_verts, obj_ids)

        return {
            'initial_smpl_pose': human_pose,
            'initial_smpl_shape': human_shape,
            'initial_obj_pose': init_obj_pose, 
            'initial_obj_trans': init_obj_trans,

            'h_contacts': h_contacts,
            'o_contacts': o_contacts,

            'smpl_verts': refined_human_verts,
            'obj_verts': refined_obj_verts,
            'obj_pose': refined_obj_pose,
            'obj_trans': refined_obj_trans,

            'init_smpl_verts': init_human_verts,
`
